### ISSUE TITLE
Rules proposal: `no-array-shift` and `no-array-unshift`

### ISSUE THREAD

**Issue by fabiospampinato** [sindresorhus/eslint-plugin-unicorn#1282]
I'd like to have rules that forbid me from using `Array.prototype.shift` and `Array.prototype.unshift` because they have truly ridiculous performance characteristics in v8 for some reason, and it's too easy to shoot myself in the foot by using these.

Also I'd imagine a lot of people aren't even aware of this, having rules for this in `eslint-plugin-unicorn` could kind of spread the word on this, which I guess would be a positive thing too.

## Fail

```js
const foo = [];
foo.shift (); // Never use shift
foo.unshift (); // Never use unshift
```

## Pass

Everything else basically, it's difficult to enumerate all the much more performant alternatives one might adopt, but general just iterating over the whole array and letting the GC collect it at the end would be much better. Or even reversing the array and pushing/popping values would be so much faster. Basically as long as the array isn't resized from its start things performance will be ok.

**Comment by fisker** [comment]
But how to achieve the same result? I don't think `array.reverse().push();array.reverse()` is going to be faster.

**Comment by fisker** [comment]
```js
const array1 = Array.from({length: 1e2}, () => Math.random())
const array2 = [...array1]

console.time('shift')
for (let i = 0; i< 1e4; i++)
    array1.shift(2)
console.timeEnd('shift')

console.time('push')
array2.reverse()
for (let i = 0; i< 1e4; i++) {
    array2.push(2)
}
array2.reverse() // even we don't reverse back, still slower
console.timeEnd('push')
```

**Comment by fabiospampinato** [comment]
> But how to achieve the same result?

It depends on what you need to do exactly, but something like this works in a lot of cases:

```js
let myArray = [...'whatever'];
myArray.forEach ( item => {
  // do something for each item from the start, without shifting
});
let myArray = []; // manually emptying it, or just let the GC reclaim it
```

Reversing the current content of the array and adding more items in reverse order to the end of it and iterating from the end should produce the same result. The difference in performance gets more significant the more items there are in the array, with just 100 items the difference in performance is less drastic. Obviously if you can avoid reversing the array that'd be better too.

**Comment by sindresorhus** [comment]
For large arrays, you should be using a queue instead if you need to "shift" a lot: https://github.com/sindresorhus/yocto-queue

**Comment by sindresorhus** [comment]
I disagree that shift/unshift is always bad. In practice, most of my usage of shift/unshift are on small arrays not in a hot-loop, so it's fine.

If we do decide to add this rule, it would be off by default.

**Comment by fisker** [comment]
I agree, `.shift()` / `.unshift()` is more readable.

**Comment by fisker** [comment]
This remind me, that I really hate `.reverse()` .

Even it's immediately reversed back after one expression or newly created array. I still avoid to use it. 

https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1039/commits/63f52efc6a7de6734204edeb638a4088ce204c77

**Comment by fabiospampinato** [comment]
> I disagree that shift/unshift is always bad. In practice, most of my usage of shift/unshift are on small arrays not in a hot-loop, so it's fine.

I agree with that, but I'd rather err on the safe side myself, trading some freedom for an assurance that things won't blow up on me.

> If we do decide to add this rule, it would be off by default.

Sounds reasonable.

**Comment by papb** [comment]
> Also I'd imagine a lot of people aren't even aware of this, having rules for this in `eslint-plugin-unicorn` could kind of spread the word on this, which I guess would be a positive thing too.

I was not aware of this... So thanks. But isn't this a V8 bug then? Is there an issue for it there?

**Comment by fabiospampinato** [comment]
> I was not aware of this... So thanks. But isn't this a V8 bug then? Is there an issue for it there?

I haven't searched the issue tracker for this, it sure sounds like a bug to me 🤷‍♂️

**Comment by fregante** [comment]
Related/preferred?

- https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1514

**Comment by fabiospampinato** [comment]
> Related/preferred?
> Rule proposal: prefer-change-array-by-copy #1514

Not for me, this proposal is about performance, copying the array is going to be fundamentally slower, especially if done for each ~shift/unshift, which would generate N-1 arrays basically.
